pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -244,3 +432,154 @@ func (r *MachineSyncReconciler) shouldMirrorCAPIMachineToMAPIMachine(ctx context
+func capiInfraMachineIsEqual(platform configv1.PlatformType, infraMachine1, infraMachine2 client.Object) (bool, error) {
+	switch platform {
+	case configv1.AWSPlatformType:
+		typedInfraMachine1, ok := infraMachine1.(*capav1beta2.AWSMachine)
+		if !ok {
+			return false, errAssertingCAPIAWSMachine
+		}
+
+		typedinfraMachine2, ok := infraMachine2.(*capav1beta2.AWSMachine)
+		if !ok {
+			return false, errAssertingCAPIAWSMachine
+		}
+
+		return reflect.DeepEqual(typedInfraMachine1.Spec, typedinfraMachine2.Spec) && objectMetaIsEqual(typedInfraMachine1.ObjectMeta, typedinfraMachine2.ObjectMeta), nil
+	case configv1.PowerVSPlatformType:
+		typedInfraMachine1, ok := infraMachine1.(*capibmv1.IBMPowerVSMachine)
+		if !ok {
+			return false, errAssertingCAPIIBMPowerVSMachine
+		}
+
+		typedinfraMachine2, ok := infraMachine2.(*capibmv1.IBMPowerVSMachine)
+		if !ok {
+			return false, errAssertingCAPIIBMPowerVSMachine
+		}
+
+		return reflect.DeepEqual(typedInfraMachine1.Spec, typedinfraMachine2.Spec) && objectMetaIsEqual(typedInfraMachine1.ObjectMeta, typedinfraMachine2.ObjectMeta), nil
+	default:
+		return false, fmt.Errorf("%w: %s", errPlatformNotSupported, platform)
+	}
+}
+
+// updateSynchronizedConditionWithPatch updates the synchronized condition
+// using a server side apply patch. We do this to force ownership of the
+// 'Synchronized' condition and 'SynchronizedGeneration'.
+func (r *MachineSyncReconciler) updateSynchronizedConditionWithPatch(ctx context.Context, mapiMachine *machinev1beta1.Machine, status corev1.ConditionStatus, reason, message string, generation *int64) error {
+	var severity machinev1beta1.ConditionSeverity
+	if status == corev1.ConditionTrue {
+		severity = machinev1beta1.ConditionSeverityNone
+	} else {
+		severity = machinev1beta1.ConditionSeverityError
+	}
+
+	conditionAc := machinev1applyconfigs.Condition().
+		WithType(consts.SynchronizedCondition).
+		WithStatus(status).
+		WithReason(reason).
+		WithMessage(message).
+		WithSeverity(severity)
+
+	setLastTransitionTime(consts.SynchronizedCondition, mapiMachine.Status.Conditions, conditionAc)
+
+	statusAc := machinev1applyconfigs.MachineStatus().
+		WithConditions(conditionAc)
+
+	if status == corev1.ConditionTrue && generation != nil {
+		statusAc = statusAc.WithSynchronizedGeneration(*generation)
+	}
+
+	msAc := machinev1applyconfigs.Machine(mapiMachine.GetName(), mapiMachine.GetNamespace()).
+		WithStatus(statusAc)
+
+	if err := r.Status().Patch(ctx, mapiMachine, util.ApplyConfigPatch(msAc), client.ForceOwnership, client.FieldOwner(controllerName)); err != nil {
+		return fmt.Errorf("failed to patch MAPI machine status with synchronized condition: %w", err)
+	}
+
+	return nil
+}
+
+// setLastTransitionTime determines if the last transition time should be set or updated for a given condition type.
+func setLastTransitionTime(condType machinev1beta1.ConditionType, conditions []machinev1beta1.Condition, conditionAc *machinev1applyconfigs.ConditionApplyConfiguration) {
```

Comment:
I reckon a bunch of these we'll eventually want in a utilities package / as helpers to be used across controllers